File: core/ImplementTriePrefixTree/ImplementTriePrefixTree.py
class TreeNode:
    def __init__(self, char=None, end=None):
        self.char = char
        self.end = end
        self.next = {}

class Trie:

    def __init__(self):
        self.root = TreeNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.next:
                node.next[ch] = TreeNode(ch, False)
            node = node.next[ch]
        node.end = True

    def search(self, word: str) -> bool:
        node = self.root
        for ch in word:
            if ch not in node.next:
                return False
            node = node.next[ch]
        return node.end

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for ch in prefix:
            if ch not in node.next:
                return False
            node = node.next[ch]
        return True

File: core/ImplementTriePrefixTree/ImplementTriePrefixTree.java
package com.dbc.code;

import java.util.HashMap;
import java.util.Map;

public class ImplementTriePrefixTree {
    class TreeNode {
        public char token;
        public boolean end;
        public final Map<Character, TreeNode> next = new HashMap<>();
        public TreeNode() {}
        public TreeNode(char ch, boolean end) {
            this.token = ch;
            this.end = end;
        }
    }

    private final TreeNode root = new TreeNode();
    public ImplementTriePrefixTree() {
    }

    public void insert(String word) {
        TreeNode node = this.root;
        for (int i = 0; i < word.length(); i++) {
            if (!node.next.containsKey(word.charAt(i))) node.next.put(word.charAt(i), new TreeNode(word.charAt(i), false));
            node = node.next.get(word.charAt(i));
        }
        node.end = true;
    }

    public boolean search(String word) {
        TreeNode node = this.root;
        for (int i = 0; i < word.length(); i++) {
            if (!node.next.containsKey(word.charAt(i))) return false;
            node = node.next.get(word.charAt(i));
        }
        return node.end;
    }

    public boolean startsWith(String prefix) {
        TreeNode node = this.root;
        for (int i = 0; i < prefix.length(); i++) {
            if (!node.next.containsKey(prefix.charAt(i))) return false;
            node = node.next.get(prefix.charAt(i));
        }
        return true;
    }
}
